crates/typst/src/foundations/calc.rs

```diff
@@ -459,6 +465,149 @@ pub fn ln(
+pub fn bit_and(
+    /// The left-hand operand of the bitwise AND.
+    lhs: i64,
+    /// The right-hand operand of the bitwise AND.
+    rhs: i64,
+) -> i64 {
+    lhs & rhs
+}
+
+#[func(title = "Bitwise OR")]
+/// Calculates the bitwise OR between two integers.
+///
+/// For the purposes of this function, the parameters are treated as signed
+/// integers of 64 bits.
+///
+/// ```example
+/// #calc.bit-or(64, 32)
+/// ```
+pub fn bit_or(
+    /// The left-hand operand of the bitwise OR.
+    lhs: i64,
+    /// The right-hand operand of the bitwise OR.
+    rhs: i64,
+) -> i64 {
+    lhs | rhs
+}
+
+#[func(title = "Bitwise XOR")]
+/// Calculates the bitwise XOR between two integers.
+///
+/// For the purposes of this function, the parameters are treated as signed
+/// integers of 64 bits.
+///
+/// ```example
+/// #calc.bit-xor(64, 96)
+/// ```
+pub fn bit_xor(
+    /// The left-hand operand of the bitwise XOR.
+    lhs: i64,
+    /// The right-hand operand of the bitwise XOR.
+    rhs: i64,
+) -> i64 {
+    lhs ^ rhs
+}
+
+#[func(title = "Bitwise Left Shift")]
+/// Shifts the operand's bits to the left by the specified amount.
+///
+/// For the purposes of this function, the operand is treated as a signed
+/// integer of 64 bits. An error will occur if the result is too large to
+/// fit in a 64-bit integer.
+///
+/// ```example
+/// #calc.bit-lshift(33, 2)
+/// #calc.bit-lshift(-1, 3)
+/// ```
+pub fn bit_lshift(
+    /// The operand whose bits will be shifted.
+    operand: i64,
+
+    /// The amount of bits to shift. Must not be negative.
+    shift: u32,
+) -> StrResult<i64> {
+    Ok(operand.checked_shl(shift).ok_or("the result is too large")?)
+}
+
+#[func(title = "Bitwise Right Shift")]
+/// Shifts the operand's bits to the right by the specified amount.
+/// Performs an arithmetic shift by default (extends the sign bit to the left,
+/// such that negative numbers stay negative), but that can be changed by the
+/// `logical` parameter.
+///
+/// For the purposes of this function, the operand is treated as a signed
+/// integer of 64 bits.
+///
+/// ```example
+/// #calc.bit-rshift(64, 2)
+/// #calc.bit-rshift(-8, 2)
+/// #calc.bit-rshift(-8, 2, logical: true)
+/// ```
+pub fn bit_rshift(
+    /// The operand whose bits will be shifted.
+    operand: i64,
+
+    /// The amount of bits to shift. Must not be negative.
+    /// This is capped at the operand's bits minus 1 (63).
+    /// Therefore, the shift will always succeed.
+    shift: u32,
+
+    /// Toggles whether a logical (unsigned) right shift should be performed
+    /// instead of arithmetic right shift (the default).
+    /// If this is `true`, negative operands will not preserve their sign bit,
+    /// and bits which appear to the left after the shift will be 0.
+    /// This parameter has no effect on non-negative operands.
+    #[named]
+    #[default(false)]
+    logical: bool,
+) -> i64 {
+    // Don't panic on excessive shift.
+    let shift = shift.min(i64::BITS - 1);
```

Comment:
IMO we should perform a checked right shift, like is already done with the left shift. This increases consistency with that function, as well as all the other operations which also perform checked operations.

Reply:
What I wanted initially was something like `saturating_shr`, but it doesn't seem to exist in Rust. So the next best thing was using a checked right shift, but to avoid a panic I'd have to provide a default value (`.unwrap_or(something)`). I realized that the default value would just be the result of capping the shift like that. So that's what I did.

Reply:
Regarding throwing an error or not, personally I think it's just not very useful to error if the shift is too large here. The result would be the same regardless.

Reply:
What I mean is that an error would be thrown when `checked_shr` fails.

Reply:
Yeah.. I just think it should be fine for Typst to be "smart" here, considering that `x >> y` where `y > 63` can be written as `(x >> y mod 63) >>^(quot(y, 63)) 63`, where `a >>^n b` indicates repeating the `a >> b` operation `n` times. (Those extra repetitions are, generally, useless though.)

The only problem with my approach, currently, is that a logical shift of `-1` will stop at `1` for `shift >= 63`, when it should stop at `0` when `shift > 63` (if we want said smart behavior above).

Reply:
There are three possible behaviors for an overflowing shift right operation:
- Wrap around, i.e. perform `operand >> (shift % 64)` (as done by JS)
- Set the result to zero or negative one, depending on signedness (as done by C)
- Error (as done by Rust)

To me, erroring seems like the most reasonable behavior, mainly because wrong assumptions will fail early.
I would be fine with the "zeroing" behavior too though, as long as it really behaves like it does in C. As it is right now, it seems incorrect as you already pointed out.

Reply:
I've fixed the code to properly saturate at 0 now, like C.

I'm thinking here that this behavior allows the right shift to function regardless of amount of bits in the internal representation of integers, so this might make more sense for Typst overall as a more high-level language. Left shift is just a bit unfortunate in this regard; same as `bit-not`.

Although maybe we could allow users to pick less bits for `bit-not`, or choose signed / unsigned? That would probably be better suited for a future PR though.